packages/store/src/actions/action-registry.ts

```diff
@@ -0,0 +1,40 @@
+import { Injectable, OnDestroy } from '@angular/core';
+import { ɵActionHandlerMetaData } from '@ngxs/store/internals';
+
+interface InvokableActionHandlerMetaData extends ɵActionHandlerMetaData {
+  path: string;
+  instance: any;
+}
+
+@Injectable({ providedIn: 'root' })
+export class NgxsActionRegistry implements OnDestroy {
+  // Instead of going over the states list every time an action is dispatched,
+  // we are constructing a map of action types to lists of action metadata.
+  // If the `@@Init` action is handled in two different states, the action
+  // metadata list will contain two objects that have the state `instance` and
+  // method names to be used as action handlers (decorated with `@Action(InitState)`).
+  private readonly _actionTypeToHandlersMap = new Map<
+    string,
+    Set<InvokableActionHandlerMetaData>
```

Comment:
This should be:
```suggestion
    Set<ActionHandlerFn>
```
Where the signature of the action handler is:
```typescript
type ActionHandlerFn = (action: Instance<ActionType>) => void | Promise<void> | Observable<unknown>;
```